### PR TITLE
Witness generation with copy constraints

### PR DESCRIPTION
**Pull request by georgwiese**
This PR adds witness generation support for copy constraints: Whenever a cell value is determined, this value is copied to all cells in its equivalence class. This allows us to do witgen for arbitrary Plonkish circuits (which would be detected as block machines) *as long as the circuit is topologically sorted* (because otherwise, our row-by-row solving strategy does not work.

Copy constraints are currently only supported in the language as `connect` identities, as opposed to lists of cell pairs that belong to the same equivalence class. Connecting this to the PIL input should be part of another PR.

### CODE REVIEW

File: executor/src/witgen/processor.rs [hunk 1]
```diff
@@ -37,6 +40,57 @@ impl<'a, T: FieldElement> OuterQuery<'a, T> {
     }
 }
 
+pub fn make_copy_constraints<T: FieldElement>(
```
Comment (georgwiese) [inline executor/src/witgen/processor.rs:43]: Tries to detect if we're in the `secondary_machine_plonk` example and hard-codes the copy constraints. We could also remove this and mark the test as `should_fail`, maybe that's better?

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Global copy constraints, represented as a graph on all cells. Each node has one outgoing edge
+/// that points to the next cell that should be equal to it, forming a set of cycles.
+pub struct CopyConstraints<CellId> {
+    /// The outgoing edge for each cell. If a cell is not in the map, it is its own next cell.
+    edges: BTreeMap<CellId, CellId>,
+}
+
+// Can't derive Default if CellId doesn't implement Default :o
```
Comment (chriseth) [inline executor/src/witgen/data_structures/copy_constraints.rs:10]:
```suggestion
```

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+impl<CellId: Ord + Copy> Default for CopyConstraints<CellId> {
+    fn default() -> Self {
+        Self {
+            edges: Default::default(),
+        }
+    }
+}
+
+impl<CellId: Ord + Copy> CopyConstraints<CellId> {
+    pub fn new(constraint_pairs: &[(CellId, CellId)]) -> Self {
```
Comment (chriseth) [inline executor/src/witgen/data_structures/copy_constraints.rs:20]: Can you add a short doc string that explains if this just adds edges to the graph or if it forces the pairs to be in the same equivalence class.

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    pub fn next(&self, cell_ref: CellId) -> CellId {
+        self.edges.get(&cell_ref).copied().unwrap_or(cell_ref)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.edges.is_empty()
+    }
+
+    fn is_connected(&self, a: CellId, b: CellId) -> bool {
+        for cell in self.iter_equivalence_class(a) {
```
Comment (chriseth) [inline executor/src/witgen/data_structures/copy_constraints.rs:56]:
```suggestion
        self.iter_equivalence_class(a).contains(b)
```

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    /// Returns the next cell in the cycle that the given cell is part of.
+    pub fn next(&self, cell_ref: CellId) -> CellId {
+        self.edges.get(&cell_ref).copied().unwrap_or(cell_ref)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.edges.is_empty()
+    }
+
+    fn is_connected(&self, a: CellId, b: CellId) -> bool {
```
Comment (chriseth) [inline executor/src/witgen/data_structures/copy_constraints.rs:55]: are_connected?

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    /// Returns an iterator over the equivalence class of the given cell. The first element is the
+    /// given cell.
+    pub fn iter_equivalence_class(&self, start: CellId) -> impl Iterator<Item = CellId> + '_ {
+        std::iter::successors(Some(start), move |&cell_ref| {
+            let next = self.next(cell_ref);
+            if next == start {
+                None
+            } else {
+                Some(next)
+            }
```
Comment (chriseth) [inline executor/src/witgen/data_structures/copy_constraints.rs:73]:
```suggestion
            (next != start).then_some(next)
```

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        })
+    }
+
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/data_structures/copy_constraints.rs:77]: Can't you use `cfg(test)` here?

File: executor/src/witgen/data_structures/copy_constraints.rs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            return;
+        }
+
+        // In the general case, a and b will already be in a cycle,
+        // where a has a next node n_a and b has a next node n_b.
+        // We want to change the edges such that a -> n_b and b -> n_a.
+        // If a node does not have an entry in the edge list, its previous
+        // and next nodes are the node itself.
+        let n_a = self.edges.get(&a).copied().unwrap_or(a);
+        let n_b = self.edges.get(&b).copied().unwrap_or(b);
```
Comment (chriseth) [inline executor/src/witgen/data_structures/copy_constraints.rs:41]:
```suggestion
        let n_a = self.next(a);
        let n_b = self.next(b);
```

File: executor/src/witgen/rows.rs [hunk 9]
```diff
@@ -58,6 +58,16 @@ impl RowIndex {
             num_rows,
         }
     }
+
+    pub fn to_local(self, row_offset: &RowIndex) -> usize {
```
Comment (chriseth) [inline executor/src/witgen/rows.rs:62]: Can you document this? What is this local to?

File: executor/src/witgen/processor.rs [hunk 10]
```diff
@@ -366,12 +426,50 @@ Known values in current row (local: {row_index}, global {global_row_index}):
         progress
     }
 
+    fn handle_copy_constraints(
```
Comment (chriseth) [inline executor/src/witgen/processor.rs:429]:
```suggestion
    fn propagate_along_copy_constraints(
```

File: executor/src/witgen/processor.rs [hunk 11]
```diff
@@ -366,12 +426,50 @@ Known values in current row (local: {row_index}, global {global_row_index}):
...
+            let row = self.row_offset + row_index + poly.next as usize;
+
+            // Have to materialize the other cells to please the borrow checker...
+            let others = self
+                .copy_constraints
+                .iter_equivalence_class((poly.poly_id, row))
+                .skip(1)
+                .collect::<Vec<_>>();
+            for (other_poly, other_row) in others {
+                let expression = &self.fixed_data.witness_cols[&other_poly].expr;
```
Comment (chriseth) [inline executor/src/witgen/processor.rs:447]: In principle, this could also be a fixed column, isn't it?

File: executor/src/witgen/processor.rs [hunk 12]
```diff
@@ -342,17 +398,21 @@ Known values in current row (local: {row_index}, global {global_row_index}):
...
+                // Build RowUpdater
+                // (a bit complicated, because we need two mutable
+                // references to elements of the same vector)
+                let (current, next) = self.data.mutable_row_pair(row_index);
+                let mut row_updater =
+                    RowUpdater::new(current, next, self.row_offset + row_index as u64);
                 row_updater.apply_update(poly, c);
                 progress = true;
+
+                if !self.copy_constraints.is_empty() {
```
Comment (chriseth) [inline executor/src/witgen/processor.rs:413]: Maybe move the early return into handle_copy_constraints?

File: executor/src/witgen/processor.rs (hunk 1 above)
Comment (chriseth) [reply executor/src/witgen/processor.rs:43]: Can you at least create an issue to replace this by properly handling the pil part?

Comment (chriseth) [reply executor/src/witgen/processor.rs]: Yes, especially when it's public.

File: executor/src/witgen/data_structures/copy_constraints.rs (hunk 3 above)
Comment (chriseth) [reply executor/src/witgen/data_structures/copy_constraints.rs:24]: I mean this is a public function, it should be documented. What is the meaning of the CellId pairs? Does this need to be a full cross product of all items in an equivalence class, or what is it?

Comment (chriseth) [reply executor/src/witgen/data_structures/copy_constraints.rs:24]: Ah read the comment. Yes, that is what I meant :)

File: executor/src/witgen/processor.rs (hunk 1 above)
Comment (georgwiese) [reply executor/src/witgen/processor.rs]: OK, removed it!

**Review by chriseth — APPROVED** [review]

**Pull request merged**